qiskit/visualization/pass_manager_visualization.py

```diff
@@ -0,0 +1,147 @@
+    """
+
+    try:
+        import pydot
+        if not HAS_GRAPHVIZ:
+            raise ImportError
+    except ImportError:
+        raise ImportError("pass_manager_drawer requires pydot and graphviz. "
+                          "Run 'pip install pydot'. "
+                          "Graphviz can be installed using 'brew install graphviz' on Mac"
+                          " or by downloading it from the website.")
+
+    passes = pass_manager.passes()
+
+    if not style:
+        style = DEFAULT_STYLE
+
+    # create the overall graph
+    graph = pydot.Dot()
+
+    # identifiers for nodes need to be unique, so assign an id
+    # can't just use python's id in case the exact same pass was
+    # appended more than once
+    node_id = 0
+
+    prev_node = None
+
+    for controller_group in passes:
+
+        # label is the name of the flow controller (without the word controller)
+        label = controller_group['type'].__name__.replace('Controller', '')
+
+        # create the subgraph for this controller
+        subgraph = pydot.Cluster(str(id(controller_group)), label=label)
+
+        for pss in controller_group['passes']:
+
+            # label is the name of the pass
+            node = pydot.Node(str(node_id),
+                              label=str(type(pss).__name__),
+                              color=_get_node_color(pss, style),
+                              shape="rectangle")
+
+            subgraph.add_node(node)
+            node_id += 1
+
+            # the arguments that were provided to the pass when it was created
+            arg_spec = inspect.getfullargspec(pss.__init__)
+            # 0 is the args, 1: to remove the self arg
+            args = arg_spec[0][1:]
+
+            num_optional = len(arg_spec[3]) if arg_spec[3] else 0
+
+            # add in the inputs to the pass
+            for arg_index, arg in enumerate(args):
+                nd_style = 'solid'
+                # any optional args are dashed
+                # the num of optional counts from the end towards the start of the list
+                if arg_index >= (len(args) - num_optional):
+                    nd_style = 'dashed'
+
+                input_node = pydot.Node(node_id, label=arg,
+                                        color="black",
+                                        shape="ellipse",
+                                        fontsize=10,
+                                        style=nd_style)
+                subgraph.add_node(input_node)
+                node_id += 1
+                subgraph.add_edge(pydot.Edge(input_node, node))
+
+            # if there is a previous node, add an edge between them
+            if prev_node:
+                subgraph.add_edge(pydot.Edge(prev_node, node))
+
+            prev_node = node
+
+        graph.add_subgraph(subgraph)
+
+    if filename:
+        # linter says this isn't a method - it is
+        graph.write_png(filename)  # pylint: disable=no-member
```

Comment:
I think it is better to fail as soon as possible, rather than do all the processing and then fail, when we could have established that it would fail earlier.

Reply:
Good point. In that case, maybe also catch `filename == None` earlier?